### PR TITLE
feat: create public squad requests mutations

### PR DESCRIPTION
**Pull request by kopancek**
## Description

Adding mutations for the parent PR:
- https://github.com/dailydotdev/daily-api/pull/1913

Unit tests added that should cover all the edge cases.

[RFC spec is here](https://dailydotdev.atlassian.net/wiki/spaces/HAN/pages/866254882/RFC-031+-+Public+Squads#Mutation---submitSquadForReview).

### RFC differences

- Instead of empty response, we return the created object on the mutation. I'm not sure how useful it is, but it was easier to test :)
- Added a query implementation, so that we can fetch the requests from the UI as well. This is important to understand, if there already are requests for review and what is the state of those.

### CODE REVIEW

File: src/schema/publicSquadRequests.ts [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+  if (lastRejected) {
+    throw new ConflictError(
+      `Public Squad request was rejected within the last ${lastDays} days!`,
+    );
+  }
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const resolvers: IResolvers<any, Context> = traceResolvers({
```
[the remark this answers is not in the document]
Comment (kopancek) [reply src/schema/publicSquadRequests.ts:193]: For sure 👍

File: src/schema/publicSquadRequests.ts [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    ): PublicSquadRequestConnection! @auth(requires: [MODERATOR])
+  }
+`;
+
+export const ensureSourceRole = async (
+  ctx: Context,
+  sourceId: string | undefined,
+  role: SourceMemberRoles = SourceMemberRoles.Admin,
+): Promise<Source> => {
+  if (sourceId) {
```
[the remark this answers is not in the document]
Comment (kopancek) [reply src/schema/publicSquadRequests.ts:141]: Yes, that's true. I'll remove this check 👍

File: src/schema/publicSquadRequests.ts [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+  }
+}
+
+const ensureNotRejected = async (
+  ctx: Context,
+  sourceId: string,
+  lastDays = REJECTED_DAYS_LIMIT,
+): Promise<void> => {
+  const daysAgo = new Date();
+  daysAgo.setDate(daysAgo.getDate() - lastDays);
```
[the remark this answers is not in the document]
Comment (kopancek) [reply src/schema/publicSquadRequests.ts:175]: I don't think I get what you mean here. [The `setDate` docs are quite clear](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/setDate#description), that the function will overflow into the correct month.

So this behavior is quite natural:
```
const daysAgo = new Date()
daysAgo.setDate(daysAgo.getDate() - 14)
// 2024-05-02T11:04:22.565Z

daysAgo.setDate(daysAgo.getDate() - 14)
// 2024-04-18T11:04:22.565Z
```

File: src/schema/publicSquadRequests.ts [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    publicSquadRequests: forwardPagination(
+      async (
+        _,
+        args: SquadRequestsArgs,
+        ctx,
+        { limit, offset },
+        info: GraphQLResolveInfo,
+      ): Promise<PaginationResponse<GQLPublicSquadRequest>> => {
+        // we need to check that the user is squad admin, moderator is not enough
+        await ensureSourceRole(ctx, args.squadId, SourceMemberRoles.Admin);
```
[the remark this answers is not in the document]
Comment (kopancek) [reply src/schema/publicSquadRequests.ts:243]: I looked at it, but for this case, it seems needlessly complicated. We just need to check the role here.

I can add a permission to admin role, e.g. `submitSquadForReview`, but then that would be only used in this case. I liked this check more 🤷

File: src/schema/publicSquadRequests.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (kopancek) [reply src/schema/publicSquadRequests.ts:193]: Now that I look at it, source schema already has 1.6K lines of code. Are you sure we want to add more to the same file? 😅 This is pretty self-encapsulated right now.

File: src/schema/publicSquadRequests.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (capJavert) [reply src/schema/publicSquadRequests.ts]: Let's leave as is.

File: src/schema/publicSquadRequests.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (capJavert) [reply src/schema/publicSquadRequests.ts:178]: All of our schemas are big either way.

File: src/schema/publicSquadRequests.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (capJavert) [reply src/schema/publicSquadRequests.ts:160]: I personally was not aware of it from top of my mind today, had to check. With integers it is just substracting and then comparing. Same thing but more clear I would say. Let's see what the rest of the team thinks, I'll respect that.

File: src/schema/publicSquadRequests.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (kopancek) [reply src/schema/publicSquadRequests.ts:178]: Ah, I misread above. `sourceRequests` actually makes sense, will do that now 👍

File: src/schema/publicSquadRequests.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (kopancek) [reply src/schema/publicSquadRequests.ts]: Changed, since it seems it would improve readability at least for you 😉

[the remark this answers is not in the document]
Comment (capJavert) [reply src/schema/publicSquadRequests.ts]: Feels like I am getting old... 😆 .. thanks!

**Review by capJavert — APPROVED** [review]
